File: gc-kettle/kettle-spring-boot-starter/src/main/java/com/gc/kettle/starter/pool/KettlePooledObjectFactory.java
package com.gc.kettle.starter.pool;

import com.gc.kettle.actuator.meta.CustomDatabaseMeta;
import com.gc.kettle.actuator.properties.DatabaseMetaProperties;
import org.apache.commons.pool2.PooledObject;
import org.apache.commons.pool2.PooledObjectFactory;
import org.apache.commons.pool2.impl.DefaultPooledObject;
import org.pentaho.di.repository.kdr.KettleDatabaseRepository;
import org.pentaho.di.repository.kdr.KettleDatabaseRepositoryMeta;

import java.util.Objects;
import java.util.Optional;

/**
 * @author shizhongming
 * 2020/7/16 8:20 下午
 */
public class KettlePooledObjectFactory implements PooledObjectFactory<KettleDatabaseRepository> {

    private final DatabaseMetaProperties properties;

    public KettlePooledObjectFactory(DatabaseMetaProperties databaseMetaProperties) {
        this.properties = databaseMetaProperties;
    }

    /**
     * 创建资源库对象
     */
    @Override
    public PooledObject<KettleDatabaseRepository> makeObject() throws Exception {
        final KettleDatabaseRepository repository = new KettleDatabaseRepository();
        // 创建连接信息
        final CustomDatabaseMeta databaseMeta = new CustomDatabaseMeta(this.properties.getName(), this.properties.getType(),
                this.properties.getAccess(), this.properties.getHost(), this.properties.getDb(),
                this.properties.getPort(), this.properties.getDbUser(), this.properties.getDbPassword());
        final KettleDatabaseRepositoryMeta kettleDatabaseRepositoryMeta = new KettleDatabaseRepositoryMeta("kettle","kettle","Transformation description",databaseMeta);
        // 初始化资源库
        repository.init(kettleDatabaseRepositoryMeta);
        //连接资源库
        repository.connect(this.properties.getResUser(),this.properties.getResPassword());
        return new DefaultPooledObject<>(repository);
    }

    /**
     * 销毁对象
     */
    @Override
    public void destroyObject(PooledObject<KettleDatabaseRepository> pooledObject) throws Exception {
        Optional.ofNullable(pooledObject.getObject())
                .ifPresent(KettleDatabaseRepository :: disconnect);
    }

    @Override
    public boolean validateObject(PooledObject<KettleDatabaseRepository> pooledObject) {
        final KettleDatabaseRepository repository = pooledObject.getObject();
        return Objects.nonNull(repository) && repository.isConnected();
    }

    @Override
    public void activateObject(PooledObject<KettleDatabaseRepository> pooledObject) throws Exception {

    }

    @Override
    public void passivateObject(PooledObject<KettleDatabaseRepository> pooledObject) throws Exception {

    }
}

File: smart-starter/kettle-spring-boot-starter/src/main/kotlin/com/smart/starter/kettle/pool/KettlePooledObjectFactory.kt
package com.smart.starter.kettle.pool

import com.smart.kettle.common.config.DatabaseMetaProperties
import com.smart.kettle.meta.CustomDatabaseMeta
import org.apache.commons.pool2.PooledObject
import org.apache.commons.pool2.PooledObjectFactory
import org.apache.commons.pool2.impl.DefaultPooledObject
import org.pentaho.di.core.KettleEnvironment
import org.pentaho.di.repository.kdr.KettleDatabaseRepository
import org.pentaho.di.repository.kdr.KettleDatabaseRepositoryMeta

/**
 * 连接池对象工厂
 * @author ming
 * 2019/7/11 下午2:21
 */
class KettlePooledObjectFactory : PooledObjectFactory<KettleDatabaseRepository> {

    private lateinit var properties: DatabaseMetaProperties

    constructor()

    constructor(properties: DatabaseMetaProperties) {
        this.properties = properties
        KettleEnvironment.init()
    }

    /**
     * 创建资源库对象
     */
    override fun makeObject(): PooledObject<KettleDatabaseRepository> {
        // 创建资源库
        val repository = KettleDatabaseRepository()
        val databaseMeta = CustomDatabaseMeta(this.properties.name, this.properties.type,
                this.properties.access, this.properties.host, this.properties.db,
                this.properties.port, this.properties.dbUser, this.properties.dbPassword)
        val kettleDatabaseRepositoryMeta = KettleDatabaseRepositoryMeta("kettle","kettle","Transformation description",databaseMeta)
        repository.init(kettleDatabaseRepositoryMeta)
        //连接资源库
        repository.connect(this.properties.resUser,this.properties.resPassword)
        return DefaultPooledObject(repository)
    }

    override fun validateObject(pooledObject: PooledObject<KettleDatabaseRepository>): Boolean {
        return pooledObject.`object`.isConnected
    }

    override fun activateObject(pooledObject: PooledObject<KettleDatabaseRepository>) {
    }

    /**
     * 销毁对象
     */
    override fun destroyObject(pooledObject: PooledObject<KettleDatabaseRepository>) {
        pooledObject.`object`?.disconnect()
    }

    override fun passivateObject(pooledObject: PooledObject<KettleDatabaseRepository>) {
    }
}